xcvm/architecture/asset-registry/global-registry.md

```diff
@@ -0,0 +1,68 @@
+# Global Asset Registry
+
+Within `XCVM` programs, assets are identified by a global identifier, which is the same within each chain. Bitcoin for example is identified by `1`.
+
+```
+AssetId ::= u64
+```
+
+The XCVM interpreter performs a look-up when executing instructions such as `Spawn` and `Transfer` to translate the local identifier to a global one. Local identifiers are chain specific, such as `address` in EVM chains. There is a two-way supported lookup:
+
+```
+AssetsById ::= AssetId -> LocalId
+IdByAssets ::= LocalId -> AssetId
+```
+
+## Querying the Registry
+
+When constructing `Call`s, we often need to convert `LocalId`s to `LocalId`s based on the `NetworkId`. Even though USDC is an erc20 token on many chains, the actual `address` is different on each chain. For this purpose, the registry also supports converting by NetworkId:
+
+```
+AssetsByIdAndNetwork ::= AssetId NetworkId -> LocalId
```

Comment:
Since there are separate registries running on each network, why do we use network id? Let's say that a token's global identifier is 1, this will be resolved in chain A as something (some address) and chain B as something else. Could you tell a use case so that I can understand?

Reply:
I'll rephrase this section, but the use case is actually:

```
Spawn XYZ [}
    Call { LocalId }
] { {AssetId}: {Amount} }
```

If we're transferring `AssetId`, to then do `Call` on the other chain, and Call is a uniswap trade call, what asset id do we need to use when constructing the `Call` on the origin chain? (The one that AssetId maps to on the destination chain).  So the contract will want to do something like:

```
/// called by a user
fn lp_on_fantom(assetId: AssetId, amount: Balance) {
    let id_on_fantom = xcvm.assets.get(Fantom::NetworkId, assetId).unwrap();
    let program = spawn
        .on(Fantom, (assetId, amount))
        .call(Univ2::lp(0x0000, id_on_fantom, amount)) // Here we cannot use AssetId, but need a LocalId
}
```

Reply:
Makes sense, I thought we were gonna use network id for transfer instruction as well.

Reply:
> Call { LocalId }

Not sure why we use local calls?  Can we specify global ids as XCVM app developers, and then interpreters maps behind the scenes?

Reply:
So the problem is that we cannot do late-bindings, as Call { LocalId } is opague. To the interpreter its just bytes and we cannot figure out what we should put there. We could change the type of a Call payload to:

```
Call ::= bytes [(Index, Value)]
Value ::= AssetId | Self
```

And we define the handling of call to be that the interpreter inserts the mapped Values at index.

Reply:
Note that encoding becomes a lot more expensive this way.

Reply:
It might actually not technically be more expensive, but we'd probably need a different codegen tool; as for example ethers.rs will not support this.